internal/orchestrator/updates.go

```diff
@@ -0,0 +1,199 @@
+
+	"github.com/nats-io/nats.go"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"go.hollow.sh/toolbox/events"
+)
+
+var (
+	updOnce             sync.Once
+	fetchEventsInterval = 1 * time.Second
+	expectedDots        = 1 // we expect keys for KV-based status updates to be facilityCode.conditionID
+	errKeyFormat        = errors.New("malformed update key")
+	errConditionID      = errors.New("bad condition uuid")
+)
+
+func (o *Orchestrator) startUpdateListener(ctx context.Context) {
+	updOnce.Do(func() {
+		if err := ctx.Err(); err != nil {
+			o.logger.WithError(err).Info("bypassing update listener start on context error")
+			return
+		}
+		if o.statusKV {
+			status.ConnectToKVStores(o.streamBroker, o.logger)
+			go o.statusKVListener(ctx)
+		} else {
+			go o.updateListener(ctx)
+		}
+	})
+}
+
+func (o *Orchestrator) statusKVListener(ctx context.Context) {
+	// start the watchers and return the associated channels
+	installWatcher, err := status.WatchFirmwareInstallStatus(ctx)
+	if err != nil {
+		o.logger.WithError(err).Fatal("unable to watch install status KV")
+	}
+	// XXX: Alloy OOB support inventoryWatcher := status.WatchInventoryStatus(ctx)
+	o.logger.Info("listening for KV updates")
+	for {
+		select {
+		case <-ctx.Done():
+			o.logger.Info("stopping KV update listener")
+			return
+			// retrieve and process events sent by controllers.
+		case kv := <-installWatcher.Updates():
+			if o.concurrencyLimit() {
+				continue
+			}
+			if kv == nil {
+				o.logger.Debug("nil kv entry")
+				continue
+			}
+			evt, err := installEventFromKV(kv)
+			if err == nil {
+				_ = o.eventHandler.UpdateCondition(ctx, evt)
+			} else {
+				o.logger.WithError(err).Warn("error creating install condition event")
+			}
+
+			/* XXX: Uncomment this for out-of-band inventory support
+			case kv := <-inventoryWatcher.Updates():
+				if o.concurrencyLimit() {
+					continue
+				}
+				evt, err := inventoryEventFromKV(kv)
+				if err == nil {
+				o.eventHandler.UpdateCondition(ctx, evt)
+			} else {
+				o.logger.WithError(err).Warn("error creating inventory condition event")
+			} */
+		}
+	}
+}
+
+type statusKey struct {
+	facility    string
+	conditionID uuid.UUID
+}
+
+// We expect keys in the format of facilityCode.uuid-as-string. If that expectation
+// is not met, it's an error and we drop the update.
+func parseStatusKVKey(key string) (*statusKey, error) {
+	if expectedDots != strings.Count(key, ".") {
+		return nil, errKeyFormat
+	}
+	elements := strings.Split(key, ".")
+	conditionID, err := uuid.Parse(elements[1])
+	if err != nil {
+		return nil, errConditionID
+	}
+	return &statusKey{
+		facility:    elements[0],
+		conditionID: conditionID,
+	}, nil
+}
+
+// installEventFromKV converts the Flasher-native StatusValue (the value from the KV) to a
+// ConditionOrchestrator-native type that ConditionOrc can more-easily use for its
+// own purposes.
+func installEventFromKV(kv nats.KeyValueEntry) (*v1types.ConditionUpdateEvent, error) {
```

Comment:
The issue here is that a controller (here Flasher) is going to define a status object that makes sense for its own internal purposes. I think this is going to be likely for any other controller we introduce here. We really only have two options:

1. Allow the controllers to define and export their own status update format and make ConditionOrc translate what it got into what it wants (to suit its specific use case).
2. Compel the controllers to write a generic status update format back to the KV (like `v1types.ConditionUpdate`). However, this assumes that ConditionOrc is the only observer of the status being relayed back by the controllers, and I think that's unnecessarily constraining from an architectural perspective. It also forces the controllers to be concerned with the kinds of information that ConditionOrc's clients want, and it couples the controller to ConditionOrc too tightly IMO.

Reply:
I was leaning towards having [StatusValue](https://github.com/metal-toolbox/flasher/blob/aaf4dfa57f872f9d8310de1e78b292e6a775a620/types/types.go#L13) moved into Conditionorc, so its available for import by Flasher and other controllers going ahead.

re 1:  I like the more stricter/explicit approach, so does sharing the `StatusValue` type help there?  
re 2: The shared types are to be moved its own package going ahead and that would allow to not have everything tightly coupled with ConditionOrc itself (if it addresses that concern)
 
 I'd really like for folks to be able to implement a controller without having to deal the Conditionorc codebase itself such that adding a new controller or a conditionKind is just a configuration change.
 
 I could be missing something obvious here, let me know

Reply:
I'm in favor of moving StatusValue into a 3rd-party library so that Controllers can define their own without having to involve ConditionOrc. Perhaps we should be more rigorous about what the required parts of a status update are, and stipulate that when a controller adds a status value to that library, it has to also add a function to deserialize its data into this generic update-thing that ConditionOrc then uses for its own purposes? I have a few ideas we can discuss as to how we might implement such a thing. 

Aside from having further discussions about this, is there anything specific you'd like me to change now, in order to land this PR?

Reply:
> Perhaps we should be more rigorous about what the required parts of a status update are, and stipulate that when a controller adds a status value to that library, it has to also add a function to deserialize its data into this generic update-thing that ConditionOrc then uses for its own purposes? I have a few ideas we can discuss as to how we might implement such a thing.
 
+1

> is there anything specific you'd like me to change now, in order to land this PR?

Lets add the `StatusValue` type in Conditionorc to begin with so the dependency on Flasher can be removed, and then going ahead we update flasher to depend on the Conditionorc `StatusValue` instead.